engine/verification/assigner/engine.go

```diff
@@ -87,12 +97,162 @@ func (e *Engine) Process(originID flow.Identifier, event interface{}) error {
 }
 
 func (e *Engine) process(originID flow.Identifier, event interface{}) error {
-	return nil
+	return fmt.Errorf("assigner engine is not supposed to invoked by process method")
+}
+
+// handleExecutionReceipt receives a receipt that appears in a finalized container block. In case this verification node
+// is staked at the reference block of this execution receipt's result, chunk assignment is done on the execution result, and
+// the assigned chunks' locators are pushed to the chunks queue, which are made available for the chunk queue consumer (i.e., the
+// fetcher engine).
+func (e *Engine) handleExecutionReceipt(receipt *flow.ExecutionReceipt, containerBlockID flow.Identifier) {
+	receiptID := receipt.ID()
+	resultID := receipt.ExecutionResult.ID()
+	referenceBlockID := receipt.ExecutionResult.BlockID
+
+	log := log.With().
+		Hex("receipt_id", logging.ID(receiptID)).
+		Hex("result_id", logging.ID(resultID)).
+		Hex("reference_block_id", logging.ID(referenceBlockID)).
+		Hex("container_block_id", logging.ID(containerBlockID)).
+		Logger()
+
+	// verification node should be staked at the reference block id.
+	ok, err := e.stakedAtBlockID(referenceBlockID)
+	if err != nil {
+		log.Error().Err(err).Msg("could not verify stake of verification node for result at reference block id")
+		return
+	}
+	if !ok {
+		log.Debug().Msg("node is not staked at reference block id, receipt is discarded")
+		return
+	}
+
+	// chunk assignment
+	chunkList, err := e.chunkAssignments(context.Background(), &receipt.ExecutionResult)
+	if err != nil {
+		log.Error().Err(err).Msg("could not determine chunk assignment")
+		return
+	}
+	log.Info().
+		Int("total_assigned_chunks", len(chunkList)).
+		Msg("chunk assignment done")
+
+	// pushes assigned chunks to
+	e.processChunks(chunkList, resultID)
+}
+
+// processChunks receives a list of chunks all belong to the same execution result. It creates a chunk locator
+// for each of those chunks and stores the chunk locator in the chunks queue.
+//
+// Note that all chunks in the input chunk list are assume to be legitimately assigned to this verification node
+// (through the chunk assigner), and all belong to the same execution result.
+//
+// Deduplication of chunk locators is delegated to the chunks queue.
+func (e *Engine) processChunks(chunkList flow.ChunkList, resultID flow.Identifier) {
+	for _, chunk := range chunkList {
+		log := e.log.With().
+			Hex("result_id", logging.ID(resultID)).
+			Hex("chunk_id", logging.ID(chunk.ID())).
+			Uint64("chunk_index", chunk.Index).Logger()
+
+		locator := &chunks.Locator{
+			ResultID: resultID,
+			Index:    chunk.Index,
+		}
+
+		// pushes chunk locator to the chunks queue
+		ok, err := e.chunksQueue.StoreChunkLocator(locator)
```

Comment:
I don't think we need two return values here, when we already have the ability to distinguish between these two cases with the `storage.ErrAlreadyExists` sentinel.

Reply:
It wasn't part of this PR, but actually, I liked the idea of abstraction, since the application layer does not require making decisions based on the returned error. In case refactoring is needed, it would be great to let me cover it in a separate PR.